actionpack/test/dispatch/middleware_stack_test.rb

```diff
@@ -213,4 +213,69 @@ def test_delete_works
   test "includes a middleware" do
     assert_equal true, @stack.include?(ActionDispatch::MiddlewareStack::Middleware.new(BarMiddleware, nil, nil))
   end
+
+  test "referencing Rack::Sendfile via #insert_before is deprecated but still valid" do
+    @stack.use ActionDispatch::MiddlewareStack::FakeSendfile
+
+    assert_deprecated(/Rack::Sendfile is removed from the default middleware stack in Rails/) do
+      @stack.insert_before(Rack::Sendfile, BazMiddleware)
+    end
+    assert_equal BazMiddleware, @stack[2].klass
+  end
+
+  test "referencing Rack::Sendfile via #insert_after is deprecated but still valid" do
+    @stack.use ActionDispatch::MiddlewareStack::FakeSendfile
+
+    assert_deprecated(/Rack::Sendfile is removed from the default middleware stack in Rails/) do
+      @stack.insert_after(Rack::Sendfile, BazMiddleware)
+    end
+    assert_equal BazMiddleware, @stack[3].klass
+  end
+
+  test "referencing Rack::Sendfile via #swap is deprecated but still valid" do
+    @stack.use ActionDispatch::MiddlewareStack::FakeSendfile
+
+    assert_equal ActionDispatch::MiddlewareStack::FakeSendfile, @stack[2].klass
+    assert_deprecated(/Rack::Sendfile is removed from the default middleware stack in Rails/) do
+      @stack.swap(Rack::Sendfile, BazMiddleware)
+    end
+    assert_equal BazMiddleware, @stack[2].klass
+  end
+
+  test "referencing Rack::Sendfile via #move_before is deprecated but still valid" do
+    @stack.use ActionDispatch::MiddlewareStack::FakeSendfile
+
+    assert_deprecated(/Rack::Sendfile is removed from the default middleware stack in Rails/) do
+      @stack.move_before(Rack::Sendfile, FooMiddleware)
+    end
+    assert_equal BarMiddleware, @stack[0].klass
+    assert_equal FooMiddleware, @stack[1].klass
+    assert_equal ActionDispatch::MiddlewareStack::FakeSendfile, @stack[2].klass
+  end
+
+  test "referencing Rack::Sendfile via #move_after is deprecated but still valid" do
+    @stack.insert_after(FooMiddleware, ActionDispatch::MiddlewareStack::FakeSendfile)
+
+    assert_deprecated(/Rack::Sendfile is removed from the default middleware stack in Rails/) do
+      @stack.move_after(Rack::Sendfile, FooMiddleware)
+    end
+    assert_equal ActionDispatch::MiddlewareStack::FakeSendfile, @stack[0].klass
+    assert_equal FooMiddleware, @stack[1].klass
+    assert_equal BarMiddleware, @stack[2].klass
+  end
+
+  test "referencing Rack::Sendfile is not deprecated when added explicitly" do
+    assert_not_deprecated do
+      @stack.use Rack::Sendfile
+      @stack.insert_before(Rack::Sendfile, BazMiddleware)
+    end
+  end
+
+  test "referencing FakeSendfile throws an error" do
+    @stack.use ActionDispatch::MiddlewareStack::FakeSendfile
+
+    assert_raises RuntimeError do
+      @stack.insert_after ActionDispatch::MiddlewareStack::FakeSendfile, BazMiddleware
+    end
```

Comment:
It seems that `ActionDispatch::MiddlewareStack::FakeSendfile` will be visible when running `rails middleware` or when inspecting `Rails.application.middleware`[1]? As such, wouldn't the normal middleware operations (insert, move, delete, etc.) all need to still function properly?

[1] railties/lib/rails/application/default_middleware_stack.rb:25

Reply:
Yes, it will be visible when running `rails middleware`. However, it would raise an exception when targeted via middleware operation, since it isn't a real middleware - `ActionDispatch::MiddlewareStack::FakeSendfile can not be referenced in middleware operations`.

Reply:
Also this fake middleware would be removed in a future version, so we don't want it to be used in operations.

Reply:
I think this gets to much of my concern. Having a middleware entry show up in `rails middleware` but then throw an exception if it's targeted via a middleware operation seems like a violation of expectations for how middleware works. This is apt to cause confusion for developers.

It's also rather awkward for anything labeled "fake" or similar to show up in the middleware stack to begin with.

To avoid such confusion and awkwardness, what do you think about removing `FakeSendfile`? Since middleware is processed upon app start, any error will show up immediately and can be quickly fixed by the app developer. This is how changes in the default middleware stack have been historically handled.

If you feel that a targeted, helpful error message is warranted, then perhaps the deprecation in `stack.rb` could be replaced with an error message instructing to change the middleware operation reference on `Rack::Sendfile` to `ActionDispatch::Static` (again, conditional upon `Rack::Sendfile` not actually being present in the middleware stack). Such an extra warning is not within historical norms, so I'll leave it to others to determine if even this much is appropriate or if it's too much.